src/server_manager/model/account_manager.ts

```diff
@@ -0,0 +1,113 @@
+// Copyright 2020 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import {LocalStorageRepository} from '../infrastructure/repository';
+
+import {Account, AccountId, DigitalOceanCredentials} from './account';
+import {CloudProviderId} from './cloud';
+import {DigitalOceanAccount} from '../web_app/digitalocean_app/digitalocean_account';
+import {DigitalOceanConnectAccountApp} from '../web_app/digitalocean_app/connect_account_app';
+
+export const ACCOUNT_MANAGER_KEY_EXTRACTOR = (entry: PersistedAccount) => entry.id;
+export const ACCOUNT_MANAGER_KEY_COMPARATOR = (k1: AccountId, k2: AccountId) =>
+    k1.cloudProviderId === k2.cloudProviderId &&
+    k1.cloudSpecificId === k2.cloudSpecificId;
+
+// TODO: Add migration from lastDOToken
+export const LEGACY_DIGITALOCEAN_ACCOUNT_ID: AccountId = {
+  cloudSpecificId: '_LEGACY_DIGITALOCEAN_ACCOUNT_ID_',
+  cloudProviderId: CloudProviderId.DigitalOcean,
+};
+
+/**
+ * A rich domain model that provides functionality to connect and manage cloud
+ * provider accounts.
+ */
+export class AccountManager {
+  // tslint:disable-next-line:no-any
+  private accountFactories: Map<CloudProviderId, AccountFactory<any>> = new Map();
```

Comment:
Whenever we use things like "manager", "factories" when you are not in the context of management or industry, we are moving away from the real world. Let's be more object oriented as in the original sense and model concepts in the real world.

We don't need an account factory. What we need are CloudProviders. A CloudProvider should be able to create/get/connect an account .

The cloud provider should be the one responsible for serializing/deserializing the account too. This way we keep the logic in one place. No need to register functions to serialize/deserialize.

As for the list method, we don't need a flat list of accounts here. List the cloud providers instead, and for each cloud provider list the accounts. This way you already get the structure you need.

Reply:
Added a SupportedClouds domain model. This provides a list of cloud providers.

This changes the architecture a bit. Instead of having one AccountManager that is responsible for managing all cloud provider accounts, the is now a Cloud domain model for each cloud provider that is responsible for its own cloud provider specific accounts.